### ISSUE TITLE
Visually mark new comments since last reload

### ISSUE THREAD

**Issue by afuna** [dreamwidth/dw-free#1233]
From http://dw-suggestions.dreamwidth.org/1304620.html:

> When you reload the page, any comments that have been posted since you last loaded the page would be marked as visually distinct and new.

PS this looks fun!

**Comment by kareila** [comment]
Given the potential overhead, should this be a paid user feature?

**Comment by rahaeli** [comment]
Depends on the implementation! If we do it browser-side, then it can probably be open to everyone. If we do it server-side, then yeah, let's start with it as a paid feature (or store "saw this" for free users for 3 days and paid users for a year, or whatever).

**Comment by woggy** [comment]
Huh. Hadn't even thought to do it browser-side (storing in the cookie, I would assume?)

That said, a multiple people have specifically requested it be browser-independent, and I agree with that approach, even if it does mean more server overhead.

**Comment by rahaeli** [comment]
Fair enough! In that case, yeah, let's make it a paid feature. (Unless you want to get really creative with the length of time stuff gets stored, like I said, but it'd probably just get frustrating for free users to constantly see NEW markers everywhere on old posts, so let's just only do it for paid users.)

**Comment by woggy** [comment]
Okay, needs some more Community Input on this.  @kareila suggested storing the relevant data as a "Last Viewed" datestamp associated with the entry.  My original thought was for a bool "is this new" associated with each individual comment. Per-comment makes it easy (possible?) to follow behavior in multiple-pages-of-comments threads, and (should) allow for a fairly seamless way to handle comments that were Previously Screened.  However, per-comment would be a *drastically* larger load on the database.  Obviously I'd like to be as light on the servers as possible, but being stumped on architecture design to allow all the functionality we want.

**Comment by rahaeli** [comment]
Hmmm. "Comment added since last you loaded the page" would be okay, but would that (for instance) mark a new comment on page 38 as read after you look at page 1, or mark a comment deep inside a nested comment thread as read if you loaded the parent page and then loaded the sub-thread? That would make this pretty useless for high-volume discussion communities like fail_fandomanon or rpanons, which is the sort of use case where this sort of feature would be most useful.

Hmmmmmm. Maybe a list per-entry that includes the # of any comment object that's been loaded-in-full on the page (so: nested comment gets added to the list when it's expanded, comment-on-page-38 gets added to the list when page 38 is viewed), and check each comment against the list when the page is loaded to decide whether to flag it or not? That would be easily memcacheable, and you'd only have to load the data once for each pageload and do the comparison on-the-fly. That might be tricky to figure out what counts as 'viewed', especially if a comment is expanded inline, but it may be the best compromise between usability and load.

@zorkian, do you have any suggestions? 

(ps: I am very excited for this feature)

**Comment by zorkian** [comment]
Can we write out a strong specification to know exactly what needs covering? I've implemented a dumb version of this 10 years ago for LJ but it never shipped. What about edits, do they show up as new? What about new comments on page you aren't on? What about if you switch in and out of flat mode? What if comments are screened when you view and later get unscreened?

If we can write out all the specifications and edge cases and desired behavior I can write up an architecture. If we want it to be totally accurate it's going to be a decent amount of load possibly.

Probably we are going to have to accept using memcache for this and saying that we will only track read state for some days. If you come back after a week we won't know what's new really.

If that's not acceptable we will certainly want to make it a paid feature. Storing this in the DB means a write on nearly every read request which is pretty heavy unless we get really really smart.

**Comment by woggy** [comment]
My take on what should be marked New (at least the points you've raised), as based on my (hazily remembered) experience with the Greasemonkey script (which, sadly, I don't have a copy of anymore to dig through their implementation details) - but I suspect that's what most users are going to treat as the Expected Behavior:

Edits: Maybe should, maybe shouldn't - I could be convinced either way. Fairly sure the script ignores edits.
Comments on Page You're Not On: absolutely treated as new.
Various view modes: I feel like any view that loads the comment in full should flag it as no-longer-new. Not sure about what the script does with collapsed comments (and my own personal use case doesn't run into them often enough to have an opinions about what Should Happen)
Screened comments: Freshly-unscreened comments should be treated as new.

There's probably other edge cases I'm not thinking of.

**Comment by kaberett** [comment]
It is probably also worth comparing with what LJ actually does these days, as that's behaviour many users are likely to expect? Which at least gives a spec to disagree with, in principle, except that unfortunately the best documentation of What It Actually Does I've found so far is http://news.livejournal.com/140969.html, which is... not super... clear.

**Comment by rahaeli** [comment]
I think @woggy is right and the new-comments Greasemonkey script is more likely to be people's point of comparison -- the big audience for this feature is the various anon discussion comms, and they've mostly moved 100% over to DW these days.

If somebody comes up with a list of "how should this work?" questions, I can post to dw-biz...

**Comment by azurelunatic** [comment]
Looks like the Greasemonkey script is this one? http://userscripts-mirror.org/scripts/review/2494 

The script stores information for the last NNNN entries viewed, rather than strictly time-bound, to keep a sensible upper limit on the log size. 

It looks at anything that has an anchor with a comment ID on the currently loaded entry-and-comments page, and anything on a lastn page that has a ?nc= link. 

From any lastn page, it tries to see if there are any new comments; this is strictly based on comment count and does not detect edits or delete-one-add-one situations. (It quietly suppresses -1 situations.) 

It tries to log collapsed comments as read, but this fails on some styles which do not expose the comment anchors for collapsed comments. So people will be used to some inconsistency with collapsed threads. 

It tries to let you jump to the next unseen comment on comments pages by giving you a helpful link after the "this is a new comment" marker. And I think you can use keyboard commands to jump around and also expand threads. 

You can mark all as read/unread. You can mark a thread as unread. Expanding a collapsed thread marks it as read.
